src/Device/Driver/FANET/Parser.cpp

```diff
@@ -0,0 +1,304 @@
+  char *ptr = strchr(cmd_str, '\r');
+  if(ptr == nullptr)
+    ptr = strchr(cmd_str, '\n');
+  if(ptr != nullptr)
+    *ptr = '\0';
+  while(*cmd_str == ' ')
+    cmd_str++;
+  return cmd_str;
+}
+
+GeoPoint
+FanetDevice::Payload2GeoPoint(const uint8_t *buf) {
+  if(buf == nullptr)
+    return GeoPoint::Invalid();
+
+  // Integer values
+  unsigned int lati = buf[2] << 16 | buf[1] << 8 | buf[0];
+  if(lati & 0x00800000)
+    lati |= 0xFF000000;
+  unsigned int loni = buf[5] << 16 | buf[4] << 8 | buf[3];
+  if(loni & 0x00800000)
+    loni |= 0xFF000000;
+
+  return GeoPoint(Angle::Degrees((float)loni / 46603.0f), Angle::Degrees((float)lati / 93206.0f));
+}
+
+float
+FanetDevice::Payload2Float(uint8_t byte, float scale) {
+  auto value = (float)(byte & 0x7F);
+  if(byte & 1 << 7)
+    return (value * scale);
+  else
+    return value;
+}
+
+bool
+FanetDevice::DecodeServicePayload(uint8_t *payload, int16_t payload_length, FanetStation *station) {
+  if(payload == nullptr || payload_length == 0)
+    return false;
+
+  uint16_t payload_pos = 1;
+
+  // Skip extended header
+  if(IsNthBitSet(payload[0], FANET_SERVICE_BIT_EHEADER))
+  {
+    // TODO: decode extended header byte if set, currently not needed/used
+    payload_pos++;
+  }
+
+  // Return if no location
+  if(payload_pos + 6 > payload_length)
+    return false;
+
+  station->location = Payload2GeoPoint(&payload[payload_pos]);
+  if( ! station->location.IsValid())
+    return false;
+  station->location_available = true;
+  payload_pos += 6;
+
+  /*
+   * Temperature (+1byte in 0.5 degree, 2-Complement)
+   */
+  if(IsNthBitSet(payload[0], FANET_SERVICE_BIT_TEMP))
+  {
+    if(payload_pos + 1 > payload_length)
+      return false;
+    station->temperature = (float)payload[payload_pos++] / 2.0f;
+  }
+
+  /*
+   * 	Wind (+3byte: 1byte Heading in 360/256 degree, 1byte speed and 1byte
+	 *  gusts in 0.2km/h (each: bit 7 scale 5x or 1x, bit 0-6))
+   */
+  if(IsNthBitSet(payload[0], FANET_SERVICE_BIT_WIND))
+  {
+    if(payload_pos + 3 > payload_length)
+      return false;
+
+    station->wind_dir_deg = (((float)payload[payload_pos++]) / 0.7083333334f);
+    station->wind_speed_kmph =
+        Payload2Float(payload[payload_pos++], 5.0f) * 0.2f;
+    station->wind_gust_kmph = Payload2Float(payload[payload_pos++], 5.0f) * 0.2f;
+  }
+
+  /*
+   * Humidity (+1byte: in 0.4% (%rh*10/4))
+   */
+  if(IsNthBitSet(payload[0], FANET_SERVICE_BIT_HUMIDITY))
+  {
+    if(payload_pos + 1 > payload_length)
+      return false;
+    station->rel_humidity = payload[payload_pos++] / 2.5f;
+  }
+
+  /*
+   * Barometric pressure normailized (+2byte: in 10Pa, offset by 430hPa,
+   * unsigned little endian (hPa-430)*10)
+   */
+  if(IsNthBitSet(payload[0], FANET_SERVICE_BIT_PRESSURE))
+  {
+    if(payload_pos + 2 > payload_length)
+      return false;
+    int lsb = payload[payload_pos++];
+    int msb = payload[payload_pos++];
+    station->preasure_hpa = (((msb << 8) | lsb) / 10.0) + 430.0;
```

Comment:
This obscure code reads an unaligned 16 bit little-endian integer from the buffer. But we have a function for that already: `ReadUnalignedLE16()`.
Also note that you're using a signed integer here; you cast the unsigned byte to a signed integer. Why the ... would you do that?